### DISCUSSION TITLE
overlay multiple datasets

### DISCUSSION THREAD

**Discussion by alexandrehoffmann** [pyvista/pyvista#2920] [Q&A]
I have two dataset I would like to overlay with different colormaps. 
Is there a simple way to do that ? Thanks in advance

Here is a minimal working example:
```py
import numpy as np
import pyvista as pv

vp = np.random.random((100, 200, 70))
wavefield = np.random.random((100, 200, 70))

# Create the spatial reference
grid = pv.UniformGrid()

# Set the grid dimensions: shape because we want to inject our values on the
#   POINT data
grid.dimensions = vp.shape

# Edit the spatial reference
grid.origin = (0.0, 0.0, 0.0)  # The bottom left corner of the data set
grid.spacing = (1, 1, 1)  # These are the cell sizes along each axis

# Add the data values to the cell data
grid.point_data["vp"] = vp.flatten(order="F")  # Flatten the array!
grid.point_data["wavefield"] = wavefield.flatten(order="F")  # Flatten the array!


# Now plot the grid!
grid.plot(show_edges=False)

```

**Comment by banesullivan** [comment]
There are a number of different ways in which you can view the two scalar fields on your mesh simultaneously. Simply viewing both with two different colormaps isn't all that nice as one will dominate the other since both scalar fields exist on exactly the same mesh geometry.

Here are a few different ways to view both fields:


```py
# View side by side
p = pv.Plotter(shape=(1,2), notebook=0)
p.add_mesh(grid, scalars='vp', cmap='viridis')
p.subplot(0,1)
# Make sure to copy the mesh!
p.add_mesh(grid.copy(), scalars='wavefield', cmap='seismic')
p.link_views()
p.view_isometric()
p.show()
```
![Screen Shot 2022-06-30 at 10 07 18 AM](https://user-images.githubusercontent.com/22067021/176726547-a9346d52-4be9-4e2b-b2c2-cd01cc6da290.png)


```py
# View one field while mapping the other to the opacity
p = pv.Plotter(notebook=0)
p.add_mesh(grid, scalars='vp', cmap='viridis', opacity='wavefield')
p.show()
```
![Screen Shot 2022-06-30 at 10 08 18 AM](https://user-images.githubusercontent.com/22067021/176726577-333fb73d-0e54-4f75-b7eb-2244d90feabd.png)

```py
# Contour one field and view cmap of the other
contours = grid.contour(scalars='vp')

p = pv.Plotter(notebook=0)
p.add_mesh(contours, scalars='wavefield', cmap='seismic')
p.show()
```
![Screen Shot 2022-06-30 at 10 12 40 AM](https://user-images.githubusercontent.com/22067021/176726646-5bcab4d4-bc7d-48a7-949f-78f04e9e42ec.png)

----

Considering random data isn't all that great, here are the same examples but with a similar dataset:

```py
import pyvista as pv
from pyvista import examples
grid = examples.download_kitchen()
```

```py
# View side by side
p = pv.Plotter(shape=(1,2), notebook=0)
p.add_mesh(grid, scalars='velocity', cmap='viridis')
p.subplot(0,1)
# Make sure to copy the mesh!
p.add_mesh(grid.copy(), scalars='p1', cmap='seismic')
p.link_views()
p.view_isometric()
p.show()
```
![Screen Shot 2022-06-30 at 10 18 27 AM](https://user-images.githubusercontent.com/22067021/176728099-785c2eab-c14e-47d2-83b4-f58f90d5f564.png)

```py
# View one field while mapping the other to the opacity
p = pv.Plotter(notebook=0)
p.add_mesh(grid.contour(), scalars='velocity', cmap='viridis', opacity='p1')
p.show()
```
![Screen Shot 2022-06-30 at 10 19 40 AM](https://user-images.githubusercontent.com/22067021/176728138-fb452336-0220-4f45-baf5-06c1f0984bcf.png)


```py
# Contour one field and view cmap of the other
contours = grid.contour(scalars='p1')

p = pv.Plotter(notebook=0)
p.add_mesh(contours, scalars='velocity', cmap='seismic')
p.show()
```
![Screen Shot 2022-06-30 at 10 20 19 AM](https://user-images.githubusercontent.com/22067021/176728166-6bb50f02-4a66-4480-8f3c-6a6e26445adf.png)

**Reply by alexandrehoffmann** [reply]
I'm sorry I did a poor job explaining what I wanted.
I want to plot something like this 
![image](https://user-images.githubusercontent.com/6158475/176732256-cdb23cb4-1df2-443f-9a96-88fdf87ebb68.png)
Where one data is ploted over the other one with transparency (and eventually a thresholding applied to the 'wavefiled'

**Reply by banesullivan** [reply]
Ugh. I wish I had a wavefield model lying around to make an actual example.


This snippet should generally work though:

```py

clipped = grid.clip_box(factor=0.5)

p = pv.Plotter(notebook=0)
p.add_mesh(clipped, scalars='vp', cmap='coolwarm')
# Make sure to copy the mesh!
p.add_mesh(clipped.copy(), scalars='wavefield', cmap='Greys_r')
p.show()
```
![Screen Shot 2022-06-30 at 11 07 29 AM](https://user-images.githubusercontent.com/22067021/176736681-b050c5ff-2e04-47f1-b6e3-682f37c3860b.png)

Depending on the values of the wavefield, you may need to add an opacity transfer function: https://docs.pyvista.org/examples/02-plot/opacity.html#transfer-functions

**Reply by alexandrehoffmann** [reply]
Just another question, 

when I apply opacity, the wavefield isn't displayed on some faces and there are some pixelization effects:
![image](https://user-images.githubusercontent.com/6158475/176840217-da0b25a9-dfe7-44fc-a5b5-89e3f1351a4a.png)
do do you think it comes from my data or is there something I can do to fix that ?

**Reply by banesullivan** [reply]
Can you share these data?

The overlapping geometries make this tough to render. This method of displaying isn't ideal.

There are a few settings we can tweak to get this looking better, but it's hard to do without the data

**Reply by alexandrehoffmann** [reply]
I don't think my bosses would be happy if I shared it :/ 
I guess we can close the thread then. 

Thanks for your advices anyway

**Reply by banesullivan** [reply]
I'll try to find some similar data to demonstrate the rendering options to improve this... hang tight
